sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,502 @@
+[section 5.7 "Optional kernel features"][1] but they also encompass the C++
+attribute `[[sycl::requires()]]` that is described in [section 5.8.1 "Kernel
+attributes"][2] and [section 5.8.2 "Device function attributes"][3].
+
+[1]: <https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:optional-kernel-features>
+[2]: <https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:kernel.attributes>
+[3]: <https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_device_function_attributes>
+
+
+## Requirements
+
+There are several categories of requirements covered by this design:
+
+* The front-end compiler must issue a diagnostic in some cases when a kernel
+  uses an optional feature.  However, the front-end compiler must **not**
+  generate a diagnostic in other cases.
+
+* The runtime must raise an exception when a kernel using optional features
+  is submitted to a device that does not support those features.  This
+  exception must be raised synchronously from the kernel invocation command
+  (e.g. `parallel_for()`).
+
+* The runtime must not raise an exception (or otherwise fail) merely due to
+  speculative compilation of a kernel for a device, when the application does
+  not specifically submit the kernel to that device.
+
+### Diagnostics from the front-end compiler
+
+By "front-end compiler", we mean the DPC++ compiler which parses DPC++ source
+code, not the JIT compiler that translates SPIR-V into native code.
+
+In general, the front-end compiler does not know which kernels the application
+will submit to which devices.  Therefore, the front-end compiler does not
+generally know which optional features a kernel can legally use.  Thus, in
+general, the front-end compiler must not issue any diagnostic simply because a
+kernel uses an optional feature.
+
+The only exception to this rule occurs when the application uses the C++
+attribute `[[sycl::requires()]]`.  When the application decorates a kernel or
+device function with this attribute, it is an assertion that the kernel or
+device function is expected to use *only* the features listed by the attribute.
+Therefore, the front-end compiler must issue a diagnostic if the kernel or
+device function uses any other optional kernel features.
+
+Note that this behavior does not change when the compiler runs in AOT mode.
+Even if the user specifies a target device via "-fsycl-targets", that does not
+necessarily mean that the user expects all the code in the application to be
+runnable on that device.  Consider an application that uses some middleware
+library, where the library's header contains kernels optimized for different
+devices.  An application should be able to compile in AOT mode with this
+library without getting errors.  Therefore the AOT compiler must not fail
+simply because the middleware header contains device code for devices that
+are not being compiled for.
```

Comment:
I agree that we need some solution for kernels that contain asm statements.  I could either include that as part of this design, or we could create a separate design for it later if we think it will be a can of worms.

How do we handle asm statements now?  Are they supported in JIT mode?  If so, it seems like we must already have a way to represent them in LLVM IR and in SPIR-V.  What does the device compiler do today with asm statements?  If there is already general way to represent asm statements in LLVM IR / SPIR-V, then it seems like the device compiler might not do anything special beyond checking their syntax.  Is that the case?

Reply:
Yes, I agree that is probably makes sense to require the user to add `[[sycl::requires()]]` to any kernel (or device function) that uses an asm statement.  To enable this, we would need to define extended aspects which correspond to each device type.  There would need to be a correspondence between each asm instruction and one of the aspect values.  I don't know the asm syntax very well.  For example, do all GEN devices accept the same asm instructions, or does each generation have its own set of instructions?  Can you propose a list of aspects that would cover the set of devices we want to support?  I can draft a DPC++ extension specification for these new aspects.  I've been thinking about doing this anyways, so now seems like a good time.

Of course, it would be more user friendly if the front-end compiler could automatically deduce the aspect requirement from the asm statement itself.  This way, the user wouldn't be required to add the `[[sycl::requires()]]` attribute manually.  That seems like it would be hard for the compiler to do, though, because it would need to recognize all possible asm instructions from all possible devices.  Am I correct in thinking that this is too difficult?

Reply:
> Am I correct in thinking that this is too difficult?

Absolutely. :-)